### PR TITLE
[Enhancement] optimize MIN/MAX on partition columns

### PR DESCRIPTION
**Pull request by murphyatwork**
## Why I'm doing:
- `SELECT MAX(partition_col)` can be used to retrieve the latest data in a table, which is very common for realtime-analytics
- But the query plan is vanilla and inefficient because it needs to scan all partitions

## What I'm doing:
Introduce several methods to optimize performance:
- Partition prune: try to scan only the last partition but not all of them
- Partition value: for List-Partition Table, we can evaluate the `MIN/MAX` in optimizer directly
- TopN: since we have TopN-Filter, the TopN is more efficient than `MIN/MAX`

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by Samrose-Ahmed** [comment]
Just curious, why is the optimization disabled if a limit is present?

**Comment by murphyatwork** [comment]
> Just curious, why is the optimization disabled if a limit is present?

good question. i didn't think about it seriously TBH.

- `SELECT min(c1) FROM tbl LIMIT 10` is meaningless actually, it's same as `SELECT min(c1) FROM tbl`. 
- `SELECT min(c1) FROM (SELECT * from tbl LIMIT 10) t`, this case is meaningful, and we should not apply this optimization on it because the subquery can be faster than a pruned query

so I can try to remove that restriction if there's no bad case.

**Review by HangyuanLiu — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PartitionColumnMinMaxRewriteRule.java [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+                        }
+                        ColumnRefOperator usedColumn =
+                                context.getColumnRefFactory().getColumnRef(usedColumns.getFirstId());
+                        Column column = scanOperator.getColRefToColumnMetaMap().get(usedColumn);
+                        if (column == null) {
+                            // not a colum on table
+                            return false;
+                        }
+                        OlapTable olapTable = (OlapTable) scanOperator.getTable();
+                        if (!olapTable.getPartitionColumns().contains(column)) {
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PartitionColumnMinMaxRewriteRule.java:107]: What about table with multiple partition columns?

Comment (murphyatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PartitionColumnMinMaxRewriteRule.java:112]: not supported

**Review by packy92 — APPROVED** [review]

**Review by HangyuanLiu — APPROVED** [review]

**Review by wyb — APPROVED** [review]

**Pull request merged**